projects/RabbitMQ.Client/client/impl/RabbitMQActivitySource.cs

```diff
@@ -0,0 +1,151 @@
+        internal const string RabbitMQDeliveryTag = "messaging.rabbitmq.delivery.tag";
+
+        internal static readonly ActivitySource s_source = new ActivitySource(Name, typeof(RabbitMQActivitySource).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion ?? "");
+        public const string Name = "RabbitMQ.Client";
+        public static bool UseRoutingKeyAsOperationName { get; set; } = true;
+        internal static bool HasListeners => s_source.HasListeners();
+
+        internal static Activity Send(string routingKey, string exchange, int bodySize)
+        {
+            if (s_source.HasListeners())
+            {
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{routingKey} publish" : "publish", ActivityKind.Producer);
+                if (activity?.IsAllDataRequested == true)
+                {
+                    PopulateMessagingTags("publish", routingKey, exchange, 0, null, bodySize, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Receive(string routingKey, string exchange, ulong deliveryTag, in ReadOnlyBasicProperties readOnlyBasicProperties, int bodySize)
+        {
+            if (s_source.HasListeners())
+            {
+                // Extract the PropagationContext of the upstream parent from the message headers.
+                DistributedContextPropagator.Current.ExtractTraceIdAndState(readOnlyBasicProperties.Headers, ExtractTraceIdAndState, out string traceparent, out string traceState);
+                ActivityContext.TryParse(traceparent, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{routingKey} receive" : "receive", ActivityKind.Consumer, parentContext);
+                if (activity != null && activity.IsAllDataRequested)
+                {
+                    PopulateMessagingTags("receive", routingKey, exchange, deliveryTag, readOnlyBasicProperties, bodySize, activity);
+                }
+
+                return activity;
+            }
+
+            return null;
+        }
+
+        internal static Activity Process(BasicDeliverEventArgs deliverEventArgs)
+        {
+            if (s_source.HasListeners())
+            {
+                // Extract the PropagationContext of the upstream parent from the message headers.
+                DistributedContextPropagator.Current.ExtractTraceIdAndState(deliverEventArgs.BasicProperties.Headers, ExtractTraceIdAndState, out string traceparent, out string traceState);
+                ActivityContext.TryParse(traceparent, traceState, out ActivityContext parentContext);
+                Activity activity = StartRabbitMQActivity(UseRoutingKeyAsOperationName ? $"{deliverEventArgs.RoutingKey} process" : "process", ActivityKind.Consumer, parentContext);
```

Comment:
based on the [OTEP-220](https://github.com/open-telemetry/oteps/blob/main/text/trace/0220-messaging-semantic-conventions-span-structure.md#general-considerations) and corresponding [spec changes about-to-be-merged](https://github.com/open-telemetry/semantic-conventions/pull/284):
- the operation name is about to be changed to `deliver` 
- context in the message SHOULD be populated as link and MAY be also a parent